src/common/menu.js

```diff
@@ -37,23 +37,37 @@ const menuData = [
   },
 ];
 
-function formatter(data, parentPath = '', parentAuthority) {
-  return data.map(item => {
-    let { path } = item;
-    if (!isUrl(path)) {
-      path = parentPath + item.path;
-    }
-    const result = {
-      ...item,
+const combinePaths = (parent, child) => `${parent.replace(/\/$/, '')}/${child.replace(/^\//, '')}`;
+
+const buildPaths = (navigation, parentPath = '') =>
+  navigation.map(route => {
+    const path = combinePaths(parentPath, route.path);
+
+    return {
+      ...route,
       path,
-      authority: item.authority || parentAuthority,
+      ...(route.routes && { routes: buildPaths(route.routes, path) }),
     };
-    if (item.routes) {
-      result.routes = formatter(item.routes, `${parentPath}${item.path}`, item.authority);
-    }
-    return result;
   });
-}
 
-const getMenuData = () => formatter(menuData);
-export default getMenuData;
+const setupParents = (routes, parentRoute = null) =>
+  routes.map(route => {
+    const withParent = {
+      ...route,
+      ...(parentRoute && { parent: parentRoute }),
+    };
+
+    return {
+      ...withParent,
+      ...(withParent.routes && {
+        routes: setupParents(withParent.routes, withParent),
+      }),
+    };
+  });
+
+const flattenRoutes = routes =>
+  routes.map(route => [route.routes ? flattenRoutes(route.routes) : [], route]).flat(Infinity);
+
+const getMenuData = () => flattenRoutes(setupParents(buildPaths(menuData)));
```

Comment:
Why do we still need this?

Reply:
Good question @portante!
umijs requires you to define your routing configuration in a config file. We do not instantiate any top-level declaration w.r.t routes as we would have done with something like  [react-router-dom](https://reactrouter.com/web/guides/quick-start).
So, we generate the breadcrumbs array by doing some post-processing based on menudata and the current location of the user. Earlier there was a lot of unnecessary post-processing which I have removed in this PR.